test/test-rivescript.coffee

```diff
@@ -832,4 +831,128 @@ exports.test_function_in_setSubroutine_return_value = (test) ->
   )
 
   bot.reply("hello", "hello person")
-  test.done()
\ No newline at end of file
+  test.done()
+
+exports.test_promises_in_objects = (test) ->
+  bot = new TestCase(test, """
+    + my name is *
+    - hello there <call>helperWithPromise <star></call> with a <call>anotherHelperWithPromise</call>
+  """)
+
+  input = "my name is Rive"
+
+  bot.rs.setSubroutine("helperWithPromise", (rs, args) ->
+    test.ok(args.length is 1)
+    test.equal(args[0], "rive")
+    return new rs.Promise((resolve, reject) -> 
+      resolve("stranger")
+    )
+  )
+
+  bot.rs.setSubroutine("anotherHelperWithPromise", (rs, args) ->
+    return new rs.Promise((resolve, reject) -> 
+      setTimeout () ->
+        resolve("delay")
+      , 1000
+    )
+  )
+
+  bot.rs.replyAsync(bot.username, input).then (reply) ->
+    test.equal(reply, "hello there stranger with a delay")
+    test.done()
+
+exports.test_replyAsync_supports_callbacks = (test) ->
+  bot = new TestCase(test, """
+    + my name is *
+    - hello there <call>asyncHelper</call>
+  """)
+
+  input = "my name is Rive"
+
+  bot.rs.setSubroutine("asyncHelper", (rs, args) ->
+    return new rs.Promise((resolve, reject) -> 
+      resolve("stranger")
+    )
+  )
+
+  bot.rs.replyAsync(bot.username, input, {}, (error, reply) -> 
```

Comment:
Is `{}` a valid value? I'd probably pass in `null` if I didn't want to pass `this`.